Tools/TestWebKitAPI/Tests/WTF/darwin/MachSendRight.cpp

```diff
@@ -0,0 +1,137 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include <wtf/MachSendRight.h>
+
+#include <mach/mach.h>
+
+namespace TestWebKitAPI {
+
+static void checkSendRefs(mach_port_t port, unsigned expected)
+{
+    unsigned count;
+    auto kr = mach_port_get_refs(mach_task_self(), port, MACH_PORT_RIGHT_SEND, &count);
+    EXPECT_EQ(kr, KERN_SUCCESS);
+    EXPECT_EQ(count, expected);
+}
+
+static mach_port_t setup()
+{
+    mach_port_t port = MACH_PORT_NULL;
+    auto kr = mach_port_allocate(mach_task_self(), MACH_PORT_RIGHT_RECEIVE, &port);
+    EXPECT_EQ(kr, KERN_SUCCESS);
+
+    kr = mach_port_insert_right(mach_task_self(), port, port, MACH_MSG_TYPE_MAKE_SEND);
+    EXPECT_EQ(kr, KERN_SUCCESS);
+    
+    checkSendRefs(port, 1);
+
+    return port;
+}
+
+TEST(MachSendRight, Adopt)
+{
+    auto port = setup();
+
+    {
+        MachSendRight right = MachSendRight::adopt(port);
+        checkSendRefs(port, 1);
+    }
+
+    checkSendRefs(port, 0);
+
+    mach_port_deallocate(mach_task_self(), port);
+}
+
+TEST(MachSendRight, Copy)
+{
+    auto port = setup();
+
+    {
+        MachSendRight right = MachSendRight::adopt(port);
+        MachSendRight copy = right;
+
+        checkSendRefs(port, 2);
+    }
+
+    mach_port_deallocate(mach_task_self(), port);
+}
+
+TEST(MachSendRight, Move)
+{
+    auto port = setup();
+
+    {
+        MachSendRight right = MachSendRight::adopt(port);
+        MachSendRight move = WTFMove(right);
+
+        checkSendRefs(port, 1);
+    }
+
+    mach_port_deallocate(mach_task_self(), port);
+}
+
+TEST(MachSendRight, Overwrite)
+{
+    auto first = setup();
+    auto second = setup();
+
+    {
+        MachSendRight firstRight = MachSendRight::adopt(first);
+        MachSendRight secondRight = MachSendRight::adopt(second);
+
+        secondRight = firstRight;
+
+        checkSendRefs(first, 2);
+        checkSendRefs(second, 0);
+    }
+
+    mach_port_deallocate(mach_task_self(), first);
+    mach_port_deallocate(mach_task_self(), second);
+}
+
+TEST(MachSendRight, OverwriteMove)
+{
+    auto first = setup();
+    auto second = setup();
+
+    {
+        MachSendRight firstRight = MachSendRight::adopt(first);
+        MachSendRight secondRight = MachSendRight::adopt(second);
+
+        secondRight = WTFMove(firstRight);
+
+        checkSendRefs(first, 1);
+        checkSendRefs(second, 0);
+    }
+
+    mach_port_deallocate(mach_task_self(), first);
+    mach_port_deallocate(mach_task_self(), second);
+}
+
```

Comment:
btw: what happens if we have a port that doesn't have a make send right?
```
mach_port_t port = MACH_PORT_NULL;
auto kr = mach_port_allocate(mach_task_self(), MACH_PORT_RIGHT_RECEIVE, &port);
auto doesThisWork = MachSendRight::adopt(port);
```

Reply:
also other codepath to test is still the case where the gives the mach send right as MACH_PORT_DEAD. (I realise this is not a MachPortRight commit but IPC::Attachment commit, just throwing it out there as an idea, no need to block on it)

Reply:
We crashed in the destructor of MachSendRight, if we'd adopted a port without a send right. I've added an extra assert to adopt(), so that this gets caught earlier.

I've also added a test case for MACH_PORT_DEAD.